### PR TITLE
[stdlib] - WIP moved aspects of ForwardIndex and BidirectionalIndex i…

### PR DESCRIPTION
**Pull request by shawnce**
@gribozavr 

Note many unit tests fail – as expected – at this point in time. I don't at the moment have any outstanding work but I was hoping to get more time over the next few days to make more progress on things. We will have to figure out how to best split up the work.

### CODE REVIEW

File: stdlib/public/core/BidirectionalCollection.swift [hunk 1]
```diff
@@ -45,7 +44,7 @@ extension BidirectionalCollection {
       return _advanceForward(i, by: n)
     }
     var i = i
-    for var offset: IndexDistance = n; offset != 0; offset = offset + 1 {
+    for _ in 0..<(-n) {
```
Comment (gribozavr) [inline stdlib/public/core/BidirectionalCollection.swift]: There's a corner case here, -IntXX.min is not representable.

File: stdlib/public/core/BidirectionalCollection.swift [hunk 2]
```diff
@@ -57,9 +56,11 @@ extension BidirectionalCollection {
       return _advanceForward(i, by: n, limit: limit)
     }
     var i = i
-    for var offset: IndexDistance = n; offset != 0 && i != limit;
-      offset = offset + 1 {
-        _previousInPlace(&i)
+    for _ in 0..<(-n) {
```
Comment (gribozavr) [inline stdlib/public/core/BidirectionalCollection.swift]: Same here.

File: stdlib/public/core/Index.swift [hunk 3]
```diff
@@ -411,6 +173,26 @@ extension RandomAccessIndex {
...
+public postfix func -- <T : BidirectionalIndex> (inout i: T) -> T {
+  fatalError("unavailable operator can't be called")
+}
+
+@available(*, unavailable, message="Please use your collections next(Index) function")
+public prefix func ++ <T : _Incrementable> (inout i: T) -> T {
+  fatalError("unavailable operator can't be called")
+}
+
+@available(*, unavailable, message="Please use your collections next(Index) function")
```
Comment (gribozavr) [inline stdlib/public/core/Index.swift]: Collection's

**Comment by gribozavr** [comment]
This generally LGTM, so I am merging in the interest of making progress together.

File: stdlib/public/core/BidirectionalCollection.swift (hunk 1 above)
Comment (shawnce) [inline stdlib/public/core/BidirectionalCollection.swift]: Yup. What you get when trying to watch a two year old while also trying to get a little programming in. (of course that would be one long negative advance to request :))

**Pull request merged**